File: Week 16/hashing.cpp
/*
 * =============================================================================
 * Week 16 — Hash Tables & Hashing  (C++17)
 * =============================================================================
 *
 * Topics covered
 * --------------
 *   1. two_sum                      — classic two-sum with unordered_map
 *   2. frequency_count              — count element frequencies
 *   3. group_anagrams               — group strings by sorted-key
 *   4. subarray_sum_equals_k        — prefix-sum + hash-map
 *   5. longest_consecutive_sequence — O(n) using unordered_set
 *
 * Complexity cheat-sheet
 * ----------------------
 *   two_sum                    O(n)  avg    |  Space O(n)
 *   frequency_count            O(n)  avg    |  Space O(n)
 *   group_anagrams             O(n * m log m) avg  |  Space O(n * m)
 *   subarray_sum_equals_k      O(n)  avg    |  Space O(n)
 *   longest_consecutive_seq    O(n)  avg    |  Space O(n)
 *
 *   (avg assumes O(1) hash lookups; worst-case can degrade to O(n) per lookup)
 *
 * Build & run
 *   g++ -std=c++17 -O2 -o hashing hashing.cpp && ./hashing
 * =============================================================================
 */

#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <cassert>
#include <sstream>
#include <optional>

// ---------------------------------------------------------------------------
// Helper — pretty-print vectors
// ---------------------------------------------------------------------------
template <typename T>
std::string vec_str(const std::vector<T>& v) {
    std::ostringstream oss;
    oss << "[";
    for (std::size_t i = 0; i < v.size(); ++i) {
        if (i) oss << ", ";
        oss << v[i];
    }
    oss << "]";
    return oss.str();
}

// ---------------------------------------------------------------------------
// 1. Two Sum — return indices of two elements that sum to target
// ---------------------------------------------------------------------------
// Complexity:  Time O(n) avg  |  Space O(n)
std::optional<std::pair<int,int>> two_sum(const std::vector<int>& nums, int target) {
    std::unordered_map<int,int> seen;   // value -> index

    for (int i = 0; i < static_cast<int>(nums.size()); ++i) {
        int complement = target - nums[i];
        if (auto it = seen.find(complement); it != seen.end()) {
            return std::make_pair(it->second, i);
        }
        seen[nums[i]] = i;
    }
    return std::nullopt;  // no solution
}

// ---------------------------------------------------------------------------
// 2. Frequency Count
// ---------------------------------------------------------------------------
// Complexity:  Time O(n) avg  |  Space O(n)
std::unordered_map<int,int> frequency_count(const std::vector<int>& nums) {
    std::unordered_map<int,int> freq;
    for (int x : nums) ++freq[x];
    return freq;
}

// ---------------------------------------------------------------------------
// 3. Group Anagrams
// ---------------------------------------------------------------------------
// Complexity:  Time O(n * m log m) avg  |  Space O(n * m)
//              n = number of strings, m = max string length
std::vector<std::vector<std::string>> group_anagrams(const std::vector<std::string>& strs) {
    std::unordered_map<std::string, std::vector<std::string>> groups;

    for (const auto& s : strs) {
        std::string key = s;
        std::sort(key.begin(), key.end());   // canonical form
        groups[key].push_back(s);
    }

    std::vector<std::vector<std::string>> result;
    result.reserve(groups.size());
    for (auto& [key, group] : groups) {
        result.push_back(std::move(group));
    }
    return result;
}

// ---------------------------------------------------------------------------
// 4. Subarray Sum Equals K — count subarrays whose sum == k
// ---------------------------------------------------------------------------
// Idea: prefix_sum[j] - prefix_sum[i] == k  =>  prefix_sum[i] == prefix_sum[j] - k
// Complexity:  Time O(n) avg  |  Space O(n)
int subarray_sum_equals_k(const std::vector<int>& nums, int k) {
    std::unordered_map<int,int> prefix_count;
    prefix_count[0] = 1;   // empty prefix

    int sum = 0, count = 0;
    for (int x : nums) {
        sum += x;
        if (auto it = prefix_count.find(sum - k); it != prefix_count.end()) {
            count += it->second;
        }
        ++prefix_count[sum];
    }
    return count;
}

// ---------------------------------------------------------------------------
// 5. Longest Consecutive Sequence
// ---------------------------------------------------------------------------
// Complexity:  Time O(n) avg  |  Space O(n)
// Each element is visited at most twice (once as a potential start, once while
// extending), giving O(n) total work.
int longest_consecutive_sequence(const std::vector<int>& nums) {
    std::unordered_set<int> num_set(nums.begin(), nums.end());

    int best = 0;
    for (int n : num_set) {
        // Only start counting from the beginning of a sequence.
        if (num_set.count(n - 1)) continue;

        int length = 1;
        int cur = n;
        while (num_set.count(cur + 1)) {
            ++cur;
            ++length;
        }
        best = std::max(best, length);
    }
    return best;
}

// ---------------------------------------------------------------------------
// main — test cases
// ---------------------------------------------------------------------------
int main() {
    std::cout << "=== Week 16: Hash Tables & Hashing ===\n\n";

    // ---- 1. Two Sum ----
    {
        std::cout << "-- Two Sum --\n";
        std::vector<int> nums = {2, 7, 11, 15};
        auto res = two_sum(nums, 9);
        assert(res.has_value());
        auto [i, j] = *res;
        assert(nums[i] + nums[j] == 9);
        std::cout << "  nums=" << vec_str(nums) << " target=9 => indices (" << i << ", " << j << ")\n";

        auto none = two_sum(nums, 100);
        assert(!none.has_value());
        std::cout << "  target=100 => no solution (correct)\n\n";
    }

    // ---- 2. Frequency Count ----
    {
        std::cout << "-- Frequency Count --\n";
        std::vector<int> nums = {1, 2, 2, 3, 3, 3};
        auto freq = frequency_count(nums);
        assert(freq[1] == 1 && freq[2] == 2 && freq[3] == 3);
        std::cout << "  {1,2,2,3,3,3} => ";
        for (auto& [k,v] : freq) std::cout << k << ":" << v << " ";
        std::cout << "\n\n";
    }

    // ---- 3. Group Anagrams ----
    {
        std::cout << "-- Group Anagrams --\n";
        std::vector<std::string> strs = {"eat", "tea", "tan", "ate", "nat", "bat"};
        auto groups = group_anagrams(strs);
        std::cout << "  input: [\"eat\",\"tea\",\"tan\",\"ate\",\"nat\",\"bat\"]\n";
        std::cout << "  groups (" << groups.size() << "):\n";
        for (const auto& g : groups) {
            std::cout << "    [";
            for (std::size_t i = 0; i < g.size(); ++i) {
                if (i) std::cout << ", ";
                std::cout << "\"" << g[i] << "\"";
            }
            std::cout << "]\n";
        }
        assert(groups.size() == 3);
        std::cout << "\n";
    }

    // ---- 4. Subarray Sum Equals K ----
    {
        std::cout << "-- Subarray Sum Equals K --\n";
        std::vector<int> nums = {1, 1, 1};
        int cnt = subarray_sum_equals_k(nums, 2);
        assert(cnt == 2);
        std::cout << "  nums=" << vec_str(nums) << " k=2 => count=" << cnt << "\n";

        std::vector<int> nums2 = {1, 2, 3};
        int cnt2 = subarray_sum_equals_k(nums2, 3);
        assert(cnt2 == 2);  // [1,2] and [3]
        std::cout << "  nums=" << vec_str(nums2) << " k=3 => count=" << cnt2 << "\n\n";
    }

    // ---- 5. Longest Consecutive Sequence ----
    {
        std::cout << "-- Longest Consecutive Sequence --\n";
        std::vector<int> nums = {100, 4, 200, 1, 3, 2};
        int len = longest_consecutive_sequence(nums);
        assert(len == 4);  // 1,2,3,4
        std::cout << "  nums=" << vec_str(nums) << " => " << len << "\n";

        std::vector<int> nums2 = {0, 3, 7, 2, 5, 8, 4, 6, 0, 1};
        int len2 = longest_consecutive_sequence(nums2);
        assert(len2 == 9);  // 0..8
        std::cout << "  nums=" << vec_str(nums2) << " => " << len2 << "\n";
    }

    std::cout << "\nAll Week 16 tests passed.\n";
    return 0;
}

File: Week 16/rust/hashing.rs
//! # Week 16: Hash Tables and Hashing
//!
//! Hash tables provide average O(1) lookup, insertion, and deletion by mapping
//! keys to array indices via a hash function.
//!
//! ## Rust's Standard Library
//! - `HashMap<K, V>` — key-value store with O(1) average operations
//! - `HashSet<T>` — set of unique values with O(1) average membership test
//!
//! Both use SipHash by default (DoS-resistant but slightly slower than
//! non-cryptographic hashes).
//!
//! ## Complexity Summary
//! | Operation                    | Time (avg) | Time (worst) | Space |
//! |-----------------------------|------------|--------------|-------|
//! | two_sum                     | O(n)       | O(n)         | O(n)  |
//! | frequency_count             | O(n)       | O(n)         | O(n)  |
//! | group_anagrams              | O(n*k log k)| O(n*k log k)| O(n*k)|
//! | subarray_sum_equals_k       | O(n)       | O(n)         | O(n)  |
//! | longest_consecutive_sequence| O(n)       | O(n)         | O(n)  |

use std::collections::{HashMap, HashSet};

// =============================================================================
// Two Sum — Classic HashMap one-pass approach
// =============================================================================

/// Given a slice of integers and a target sum, returns the indices of two
/// numbers that add up to the target.
///
/// # Algorithm
/// For each element, check if `target - element` is already in the map.
/// If yes, we found our pair. Otherwise, store `(element, index)`.
///
/// # Complexity
/// - Time: O(n) — single pass with O(1) hash lookups
/// - Space: O(n) — at most n entries in the map
///
/// # Ownership Note
/// Takes `&[i32]` (a shared slice reference) — we only need to read the data.
/// Returning `Option<(usize, usize)>` avoids needing to allocate.
fn two_sum(nums: &[i32], target: i32) -> Option<(usize, usize)> {
    let mut seen: HashMap<i32, usize> = HashMap::new();

    for (i, &num) in nums.iter().enumerate() {
        let complement = target - num;
        if let Some(&j) = seen.get(&complement) {
            return Some((j, i));
        }
        seen.insert(num, i);
    }

    None
}

// =============================================================================
// Frequency Count
// =============================================================================

/// Counts the frequency of each element in a slice.
///
/// # Complexity
/// - Time: O(n)
/// - Space: O(k) where k is the number of distinct elements
///
/// # Rust Idiom
/// The `entry` API avoids double lookups: `entry(key).or_insert(0)` returns
/// a mutable reference to the value, inserting a default if the key is absent.
fn frequency_count(nums: &[i32]) -> HashMap<i32, usize> {
    let mut freq: HashMap<i32, usize> = HashMap::new();
    for &num in nums {
        *freq.entry(num).or_insert(0) += 1;
    }
    freq
}

// =============================================================================
// Group Anagrams
// =============================================================================

/// Groups strings that are anagrams of each other.
///
/// # Algorithm
/// Two strings are anagrams if they have the same characters in sorted order.
/// Use the sorted string as a HashMap key to group anagrams together.
///
/// # Complexity
/// - Time: O(n * k log k) where n = number of strings, k = max string length
/// - Space: O(n * k)
///
/// # Ownership Note
/// Takes `&[String]` — borrows the input strings. We clone strings into the
/// result groups since we cannot move data out of a shared reference.
fn group_anagrams(strs: &[String]) -> Vec<Vec<String>> {
    let mut groups: HashMap<String, Vec<String>> = HashMap::new();

    for s in strs {
        // Sort the characters to create a canonical key
        let mut chars: Vec<char> = s.chars().collect();
        chars.sort_unstable();
        let key: String = chars.into_iter().collect();

        groups.entry(key).or_insert_with(Vec::new).push(s.clone());
    }

    // Collect all groups into a Vec (order is not guaranteed)
    groups.into_values().collect()
}

// =============================================================================
// Subarray Sum Equals K — Prefix sum + HashMap
// =============================================================================

/// Counts the number of contiguous subarrays whose elements sum to `k`.
///
/// # Algorithm
/// Maintain a running prefix sum. At each index, the number of subarrays
/// ending here with sum `k` equals the number of times `prefix_sum - k`
/// has appeared as a previous prefix sum.
///
/// Key insight: if `prefix[j] - prefix[i] == k`, then `sum(arr[i+1..=j]) == k`.
///
/// # Complexity
/// - Time: O(n) — single pass
/// - Space: O(n) — prefix sum counts in the map
fn subarray_sum_equals_k(nums: &[i32], k: i32) -> i32 {
    let mut count = 0;
    let mut prefix_sum = 0;
    // Map from prefix_sum value -> number of times it has occurred
    let mut prefix_counts: HashMap<i32, i32> = HashMap::new();

    // Base case: a prefix sum of 0 has occurred once (empty prefix)
    prefix_counts.insert(0, 1);

    for &num in nums {
        prefix_sum += num;

        // How many previous prefix sums equal (prefix_sum - k)?
        // Each one represents a subarray summing to k.
        if let Some(&c) = prefix_counts.get(&(prefix_sum - k)) {
            count += c;
        }

        *prefix_counts.entry(prefix_sum).or_insert(0) += 1;
    }

    count
}

// =============================================================================
// Longest Consecutive Sequence — HashSet approach
// =============================================================================

/// Finds the length of the longest consecutive element sequence.
///
/// # Algorithm
/// 1. Insert all numbers into a HashSet for O(1) lookups.
/// 2. For each number, check if it is the START of a sequence (i.e., `num - 1`
///    is NOT in the set). If so, count how far the sequence extends.
/// 3. This ensures each element is visited at most twice total → O(n).
///
/// # Complexity
/// - Time: O(n) — although there's a nested loop, each element is part of
///   exactly one sequence and visited at most twice across all iterations
/// - Space: O(n) — for the HashSet
fn longest_consecutive_sequence(nums: &[i32]) -> i32 {
    let num_set: HashSet<i32> = nums.iter().copied().collect();
    let mut best = 0;

    for &num in &num_set {
        // Only start counting from the beginning of a sequence
        if !num_set.contains(&(num - 1)) {
            let mut current = num;
            let mut length = 1;

            while num_set.contains(&(current + 1)) {
                current += 1;
                length += 1;
            }

            best = best.max(length);
        }
    }

    best
}

// =============================================================================
// Main — Test cases
// =============================================================================

fn main() {
    println!("=== Week 16: Hashing ===\n");

    // --- Two Sum ---
    println!("--- Two Sum ---");
    let nums = vec![2, 7, 11, 15];
    let result = two_sum(&nums, 9);
    println!("two_sum([2,7,11,15], 9) = {:?}", result);
    assert_eq!(result, Some((0, 1)));

    let result2 = two_sum(&[3, 2, 4], 6);
    println!("two_sum([3,2,4], 6) = {:?}", result2);
    assert_eq!(result2, Some((1, 2)));

    assert_eq!(two_sum(&[1, 2, 3], 10), None);
    println!("PASS: Two sum works correctly\n");

    // --- Frequency Count ---
    println!("--- Frequency Count ---");
    let freq = frequency_count(&[1, 2, 2, 3, 3, 3, 4, 4, 4, 4]);
    println!("Frequencies: {:?}", freq);
    assert_eq!(freq[&1], 1);
    assert_eq!(freq[&2], 2);
    assert_eq!(freq[&3], 3);
    assert_eq!(freq[&4], 4);
    println!("PASS: Frequency count works correctly\n");

    // --- Group Anagrams ---
    println!("--- Group Anagrams ---");
    let words: Vec<String> = vec!["eat", "tea", "tan", "ate", "nat", "bat"]
        .into_iter()
        .map(String::from)
        .collect();
    let mut groups = group_anagrams(&words);
    // Sort each group and the outer vec for deterministic comparison
    for g in &mut groups {
        g.sort();
    }
    groups.sort_by(|a, b| a[0].cmp(&b[0]));
    println!("Grouped anagrams: {:?}", groups);
    assert_eq!(groups.len(), 3);
    println!("PASS: Group anagrams works correctly\n");

    // --- Subarray Sum Equals K ---
    println!("--- Subarray Sum Equals K ---");
    let count = subarray_sum_equals_k(&[1, 1, 1], 2);
    println!("subarray_sum_equals_k([1,1,1], 2) = {}", count);
    assert_eq!(count, 2);

    let count2 = subarray_sum_equals_k(&[1, 2, 3], 3);
    println!("subarray_sum_equals_k([1,2,3], 3) = {}", count2);
    assert_eq!(count2, 2); // [1,2] and [3]

    let count3 = subarray_sum_equals_k(&[1, -1, 0], 0);
    println!("subarray_sum_equals_k([1,-1,0], 0) = {}", count3);
    assert_eq!(count3, 3); // [1,-1], [1,-1,0], [0]
    println!("PASS: Subarray sum equals k works correctly\n");

    // --- Longest Consecutive Sequence ---
    println!("--- Longest Consecutive Sequence ---");
    let result = longest_consecutive_sequence(&[100, 4, 200, 1, 3, 2]);
    println!("longest_consecutive([100,4,200,1,3,2]) = {}", result);
    assert_eq!(result, 4); // Sequence: 1,2,3,4

    let result2 = longest_consecutive_sequence(&[0, 3, 7, 2, 5, 8, 4, 6, 0, 1]);
    println!("longest_consecutive([0,3,7,2,5,8,4,6,0,1]) = {}", result2);
    assert_eq!(result2, 9); // Sequence: 0-8

    let result3 = longest_consecutive_sequence(&[]);
    assert_eq!(result3, 0);
    println!("PASS: Longest consecutive sequence works correctly\n");

    println!("All Week 16 tests passed!");
}
